### ISSUE TITLE
Exchanging order of  an "only parsing" and a "only printing" notation breaks printing.

### ISSUE THREAD

**Issue by Janno** [coq/coq#9682]
#### Description of the problem

In the following example, exchanging the order of the last two notations breaks the printing-only notation, i.e. the `Check` invocation would print `consumer..` instead of `mmatch`.
```coq
Declare Scope blafu.
Delimit Scope blafu with B.
Axiom DoesNotMatch : Type.
Axiom consumer : forall {A} (B : A -> Type) (E:Type) (x : A) (ls : list nat), unit.

Notation "|  p1  |  ..  |  pn" := (@cons _ p1 .. (@cons _ pn nil) ..) (at level 91) : blafu.
Notation "'mmatch_do_not_write' x 'in' T 'as' y 'return' 'M' p 'with_do_not_write' ls" :=
    (@consumer _ (fun y : T => p%type) DoesNotMatch x ls%B)
      (at level 200, ls at level 91, only parsing).
Notation "'mmatch' x 'in' T 'as' y 'return' 'M' p 'with' ls 'end'" :=
  (@consumer _ (fun y : T => p%type) DoesNotMatch x ls%B)
    (at level 200, ls at level 91, p at level 10, only printing,
     format "'[  ' mmatch  '/' x ']'  '/' '[  ' in  '/' T ']'  '/' '[  ' as  '/' y ']'  '/' '[  ' return  M  p ']'  with  '//' '[' ls ']'  '//' end"
     ).
Notation "'mmatch' x 'in' T 'as' y 'return' 'M' p 'with' ls 'end'" :=
  (mmatch_do_not_write x in T as y return M p with_do_not_write ls)
    (at level 200, ls at level 91, p at level 10, only parsing).
Check (mmatch 1 +2 +3+4+5+6 in nat as x return M (x = x) with | 1 end).
```

#### Coq Version
master

**Comment by staffehn** [comment]
Reduced example:
```coq
Notation "#" := I (at level 0, only parsing).
Notation "#" := I (at level 0, only printing).
Check #.
(*
I
     : True
*)

Notation "##" := I (at level 0, only printing).
Notation "##" := I (at level 0, only parsing).
Check ##.
(*
##
     : True
*)
```

**Comment by herbelin** [comment]
@Janno (short answer):

There is an underlying design choice issue (and a change since 8.9). 

Should a notation (in the sense of a string such as `"##"` or `"|  p1  |  ..  |  pn"`):
- come with a data telling if it is only printing, only parsing, or both, making it impossible to be used with a different interpretation or different levels in parsing and printing (this was the view up to 8.8 which makes the first `Check #` above failing in 8.8)
- or, should such a notation string be treated differently for parsing and printing (knowing that 8.9 partially starts to behave like this, but incompletely).

In the second case (answering also to your gitter message), `level_of_notation` would typically be duplicated into `parsing_level_of_notation` and `printing_level_of_notation` and all (re)declarations behave independently for parsing and printing.

**Comment by herbelin** [comment]
A "new" model where notations can explicitly live independently as only-parsing, or only-printing, or combined-parsing-and-printing has been implemented in #12950. This fixes the issue.

Note: there was a spurious warning for the first example which is fixed in #12949.
